keps/sig-network/20200127-disable-service-type-lb-nodeports.md

```diff
@@ -0,0 +1,167 @@
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+As of today, a Kubernetes Service with Type=LoadBalancer will always allocate a node port for every service port. Though most implementations of Service Type=LoadBalancer do require node ports, there are several implementations that do not. This KEP proposes to add a new flag on the apiserver (`--disable-loadbalancer-node-ports`) to optionally disable node port allocation for Service Type=LoadBalancer.
+
+## Motivation
+
+There are several implementations of Service Type=LoadBalancer API today that do not consume the node ports automatically allocated by Kubernetes. Some examples include but are not limited to:
+* [MetalLB](https://github.com/danderson/metallb)
+* [kube-router](https://github.com/cloudnativelabs/kube-router)
+
+The drawbacks of allocating unused node ports are that:
+* the number of load balancers is now limited to the number of available node ports
+* node ports are exposed for an LB even though they are not used -- unecessarily exposed ports can fail regulatory/compliance requirements
+
+For clusters that have integrations for Service Type=LoadBalancer but don't require node ports should have the option to disable node port allocation.
+
+### Goals
+
+* disable node port for Service Type=LoadBalancer
+* don't change any existing default behavior for Service Type=LoadBalancer
+* Service LoadBalancer node ports can safely be re-enabled in a cluster
+
+## Proposal
+
+A flag `--disable-loadbalancer-node-ports` will be added to the apiserver to disable node port allocations for Service Type=LoadBalancer. The flag's default value is `false` to ensure existing behavior is preserved.
+
+If `--disable-loadbalancer-node-ports` is `true`, the following changes are expected on the apiserver:
+- any new Service LoadBalancers will not have node ports allocated
+- existing Service LoadBalancers with node ports will continue to operate with node ports
+- new Service LoadBalancers can still get node ports assigned manually
+
+The service port allocator controller will be used to rectify the following edge cases:
+- Service LoadBalancer node port allocation is disabled and then re-enabled -- any new Services that weren't allocated node ports should get node ports assigned by the port allocator controller.
+
+## Design Details
+
+### Global API Server Flag
+
+A global apiserver flag `--disable-loadbalancer-node-ports` will be used to toggle this behavior as opposed to per-Service annotations or webhooks since Service LoadBalancer implementations are also done at a global level (via the cloud provider).
```

Comment:
This isn't *quite* true.  This seems more appropriate for each SLB implementation - and for instance metal could have multiple SLBs cooperating.

Reply:
Good point -- but even with multiple SLB implementations, the node port toggle would still have to global because:
* if at least 1 SLB depends on node ports, you enable it for all Services
* if no SLB depends on node ports, it should be disabled for all Services.
 
Is it common for an SLB implementation to only operate on a subset of service type=lb?

Reply:
Just as this subject is getting into my radar, we've seen an issue of a user trying to use this scenario of multiple LBs here: https://github.com/kubernetes/kubernetes/issues/91306

I think we might have some cases where also each LB works differently, so IMO using an attribute for the service instead of a global flag makes more sense, also for those that want to opt-in or opt-out of this behavior.

Reply:
An attribute is fine but the **default** should be sttable with an option, like; `--default-lb-mode=disable-nodeport`

Reply:
> I think we might have some cases where also each LB works differently, so IMO using an attribute for the service instead of a global flag makes more sense, also for those that want to opt-in or opt-out of this behavior.

Not sure I agree with this yet, SLB is tightly tied to the cloud providers at the moment and so far it's been a "one implementation" per cluster kind of thing, so a global cluster config make sense to me.

It sounds like we could maybe benefit from a service class annotation or type, but I don't think @thockin is going to buy that :P